src/darwin/Framework/CHIP/MTRDeviceControllerLocalTestStorage.m

```diff
@@ -0,0 +1,97 @@
+//
+/**
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTRDeviceControllerLocalTestStorage.h"
+
+#if MTR_PER_CONTROLLER_STORAGE_ENABLED
+
+static NSString * const kLocalTestUserDefaultDomain = @"org.csa-iot.matter.darwintest";
+static NSString * const kLocalTestUserDefaultEnabledKey = @"enableTestStorage";
+
+@implementation MTRDeviceControllerLocalTestStorage {
+    id<MTRDeviceControllerStorageDelegate> _passThroughStorage;
+}
+
++ (BOOL)localTestStorageEnabled
+{
+    NSUserDefaults * defaults = [[NSUserDefaults alloc] initWithSuiteName:kLocalTestUserDefaultDomain];
+    return [defaults boolForKey:kLocalTestUserDefaultEnabledKey];
+}
+
++ (void)setLocalTestStorageEnabled:(BOOL)localTestStorageEnabled
+{
+    NSUserDefaults * defaults = [[NSUserDefaults alloc] initWithSuiteName:kLocalTestUserDefaultDomain];
+    [defaults setBool:localTestStorageEnabled forKey:kLocalTestUserDefaultEnabledKey];
+}
+
+- (instancetype)initWithPassThroughStorage:(id<MTRDeviceControllerStorageDelegate>)passThroughStorage
+{
+    if (self = [super init]) {
+        _passThroughStorage = passThroughStorage;
+    }
+    return self;
+}
+
+- (nullable id<NSSecureCoding>)controller:(MTRDeviceController *)controller
+                              valueForKey:(NSString *)key
+                            securityLevel:(MTRStorageSecurityLevel)securityLevel
+                              sharingType:(MTRStorageSharingType)sharingType
+{
+    if (sharingType == MTRStorageSharingTypeNotShared) {
+        NSUserDefaults * defaults = [[NSUserDefaults alloc] initWithSuiteName:kLocalTestUserDefaultDomain];
+        NSData * storedData = [defaults dataForKey:key];
```

Comment:
As we figured out on Slack: this is going to cause anything that involves multiple controllers to break, because they will stomp on each other's storage....  Probably need to include the controller id in the key, but then the persistence aspect might break unless people use stable ids.